### PR TITLE
storage: add a ResourceVersion field for lists

### PR DESCRIPTION
**Pull request by nicks**
this allows List/Watch to better synchronize on the latest changes

fixes https://github.com/tilt-dev/tilt-apiserver/issues/67

### CODE REVIEW

**Comment by nicks** [comment]
i forget why we didn't thread all the mutexes thru RealFS...do you remember?

**Comment by milas** [comment]
I think this comment is the key:
```
	// TODO(milas): use storageVersion to ensure we don't perform stale writes
	// 	(currently, this isn't a critical priority as our use cases that rely
	// 	on RealFS do not have simultaneous writers)
```

The `MemoryFS` write function is substantially more complex and handles optimistic concurrency while the filesystem implementation...does not.

That is, I don't think there will be much meaningful difference between the current implementation and one with a mutex. Even with mutex, we can still have multiple process issues like tilt-dev/tilt#6244 [I haven't actually looked into that deeply, so it might be something else, but sounds sus].) In theory, we probably need some sort of lockfile for the resource, and once the lock is held, read back the version on-disk to determine if the write can proceed (based on resource versions).

tl;dr That resource version being storage version+1 was basically a hack to say "last one in always wins" IIRC

**Comment by nicks** [comment]
updated the PR!

ya, the problem i ran into isn't about handling optimistic writes. it's more about handling a simultaneous List() and Write(), which can definitely happen where we currently use this code. Now that List objects have a revision number, we need to be careful about returning a List() with a bad rev number.

**Comment by nicks** [comment]
(that said, the approach i added is still very naive)

**Comment by karolz-ms** [comment]
@nicks couple questions:

1. Is this change going to also add resource version for List request results for resources backed by memory storage?
2. Is there a timeline for having this change merged?

Thanks!

**Review by milas — APPROVED** [review]

**Comment by nicks** [comment]
@karolz-ms ya it should fix both file and in-memory storage

**Comment by karolz-ms** [comment]
Excellent, thanks Nick. Should I expect 0.11 release of API server soon?

**Comment by nicks** [comment]
@karolz-ms released!
